cargo-quickinstall/src/install_error.rs

```diff
@@ -107,3 +107,37 @@ impl From<JsonExtError> for InstallError {
         InstallError::JsonErr(err)
     }
 }
+
+#[derive(Debug)]
+pub enum InstallSuccess {
+    InstalledFromTarball,
+    BuiltFromSource,
+}
+
+/**
+ * Returns a status string for reporting to our stats server.
+ *
+ * The return type is a static string to encourage us to keep the cardinality vaguely small-ish,
+ * and avoid us accidentally dumping personally identifiable information into influxdb.
+ *
+ * If we find ourselves getting a lot of a particular genre of error, we can always make a new
+ * release to split things out a bit more.
+ *
+ * There is no requirement for cargo-quickinstall and cargo-binstall to agree on the status strings,
+ * but it is probably a good idea to keep at least the first two in sync.
+ */
+pub fn install_result_to_status_str(result: &Result<InstallSuccess, InstallError>) -> &'static str {
+    match result {
+        Ok(InstallSuccess::InstalledFromTarball) => "installed-from-tarball",
+        Ok(InstallSuccess::BuiltFromSource) => "built-from-source",
+        Err(InstallError::MissingCrateNameArgument(_)) => "missing-crate-name-argument",
+        Err(InstallError::CommandFailed(_)) => "command-failed",
+        Err(InstallError::IoError(_)) => "io-error",
+        Err(InstallError::CargoInstallFailed) => "cargo-install-failed",
+        Err(InstallError::CrateDoesNotExist { .. }) => "crate-does-not-exist",
+        Err(InstallError::NoFallback(_)) => "no-fallback",
+        Err(InstallError::InvalidJson { .. }) => "invalid-json",
+        Err(InstallError::JsonErr(_)) => "json-err",
+        Err(InstallError::FailToParseRustcOutput { .. }) => "fail-to-parse-rustc-output",
```

Comment:
I knew I wanted the first two, and then I also wanted a separate status for NoFallback, and adding the rest was basically free at that point.

If you want to lump "everything else" together in sql, I think you'd do something like `status is not null and status not in ('installed-from-tarball', 'built-from-source'), so the cardinality doesn't really matter that much.

Are we expecting people to get upset about this level of detail?

Reply:
Well, I think most people would like it to have minimum information collected.

For the purpose of our stats server, I think we'd only need to report, if we doesn't have an existing pre-built binary.

That would reduce information collected, while also reduce loads on our stats server, since we are on free-tier that could be rate limited at any time.

Reply:
I have a feeling that fly.io is more generous than vercel (or that our implementation is doing less work). I have had vercel forwarding all requests to fly.io since March 2023 and not had any complaints from fly.io.

If we get rate limited, it's typically from people using us in CI, who use the latest version.

Specific Example: when you made the release of cargo-binstall 1.10.4 at 23:57 UK time, within an hour, traffic for 1.10.3 had gone through the floor (with only a little spike over the next few days).

![Screenshot 2024-09-09 at 16 10 13](https://github.com/user-attachments/assets/9ee94145-e68d-4f13-9696-56147bcb14a5)

I think I would prefer to have these success numbers back, because it lets us know when we can stop supporting old clients completely (and we've demonstrated how quickly we can turn it off if we need to). It's also good for ego-stroking, which I think is quite important :-).

Reply:
Sure, but I think it might worth simplifying them a bit.

There's a lot more error variants in cargo-binstall and I don't want to report everyone of them, almost too many to report.

Reply:
BTW latest version of cargo-binstall only reports on failure due to previous payment needed error.

Reply:
Oh you already simplified it, sorry I didn't see that, approved 😀

Reply:
I have pulled out 2 of the error variants that I think are interesting, and lumped the other errors together into `other-error`

Reply:
Do you want to do the binstall change, or should I have a crack at it?

Reply:
I'm free after 10-12h, if you want to have a crack, here's the file you need to edit 

https://github.com/cargo-bins/cargo-binstall/blob/a88335d05bc9b62a204cddffb309bfc8e9fa9105/crates/binstalk-fetchers/src/quickinstall.rs#L283